### PR TITLE
#170947655 Register a new user

### PR DESCRIPTION
**Pull request by Niyonsengaeric**
#### What does this PR do?
 This PR ensures that the user creates an account and receive a JSON Web Token on success.
#### Description of Task to be completed?
- Users once registered via the registration endpoint, should receive a JSON Web Token to be required on all subsequent calls to all other endpoints that require authentication.
- test the endpoint
- Document feature using swagger

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `git checkout ft-user-registration-170947655` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database.
- create a .env file then add environment variable as shown below:
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
```
- run `npm run createTables` to create the database tables
- run `npm start` to start the server.
- Open postman and copy this URL POST `localhost:4000/api/v1/auth/signup`
- Add a comment body in the format shown below:
```
{
	"firstName": "Niyonsenga",
	"lastName": "Eric",
	"gender": "Male",
	"PassportNumber": "12345675312",
	"email": "example@gmail.com ",
	"password": "Niyonkuru@1"
}
```
A new user will be created and details will be returned.
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#170947655](https://www.pivotaltracker.com/story/show/170947655)

#### Screenshots

     Registration Successful

Screen Shot 2020-02-10 at 17 12 30

### CODE REVIEW

**Comment by Moise1** [comment]
Hi Eric! Do you mind writing JWT in full to make it easily understandable to the reader and save them the time of trying to guess the meaning? Thanks.

File: src/tests/mockData.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    lastName: 'Eric',
+    gender: 'MALE',
+    PassportNumber: '1234544546464646',
+    email: 'niyeric1@gmail.com',
+    password: 'password',
+    confirm: '12345six',
+  },
+
+};
+export default mockDate;
```
Comment (Moise1) [inline src/tests/mockData.js]: Hi Eric! You can also change this object's name from mockDate to mockData

File: src/db/migrations/20200208090739-create-user.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('Users');
+  }
+};
```
Comment (Kagorora) [inline src/db/migrations/20200208090739-create-user.js]: Please put a new line at the end of this file

File: src/db/migrations/20200208090739-create-user.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        type: Sequelize.STRING
+      },
+      gender: {
+        type: Sequelize.STRING
+      },
+      Passport: {
+        type: Sequelize.STRING
+      },
+      email: {
+        type: Sequelize.STRING
```
Comment (Kagorora) [inline src/db/migrations/20200208090739-create-user.js]: Please put a trailing comma at each line

File: src/db/models/user.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+module.exports = (sequelize, DataTypes) => {
+  const User = sequelize.define('User', {
+    firstName: DataTypes.STRING,
+    lastName: DataTypes.STRING,
+    gender: DataTypes.STRING,
+    Passport: DataTypes.STRING,
+    email: DataTypes.STRING,
+    password: DataTypes.STRING
+  }, {});
+  User.associate = function(models) {
```
Comment (Kagorora) [inline src/db/models/user.js]: Please put new line

File: src/tests/user.test.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import chai, { expect } from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../app';
+import mockData from './mockData';
+
+chai.use(chaiHttp);
+chai.should();
+
+describe('Create a user account.(POST) ', () => {
+  it('New user, it should return 201 on successful signup', (done) => {
```
Comment (Kagorora) [inline src/tests/user.test.js]: Please remove "New user", as it may not be needed

File: src/db/migrations/20200208090739-create-user.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    firstName: {
+      type: Sequelize.STRING,
+    },
+    lastName: {
+      type: Sequelize.STRING,
+    },
+    gender: {
+      type: Sequelize.STRING,
+    },
+    email: {
```
Comment (EugeneMunya) [inline src/db/migrations/20200208090739-create-user.js:19]: Can you make an email unique?

**Review by Moise1 — APPROVED** [review]

**Comment by EugeneMunya** [comment]
> #### What does this PR do?
> This PR ensures that the user creates an account and receive a JSON Web Token on success.
User will receive a token for using what?  explain what token will be used for

**Review by EugeneMunya — APPROVED** [review]

**Comment by MuhanguziDavid** [comment]
Please, add a step in your PR description about running migrations and any environment variables if necessary.

**Comment by MuhanguziDavid** [comment]
Ensure that `PassportNumber` starts with a lower case letter like the rest of the variables (for purposes of uniformity). 
Screenshot 2020-02-10 at 17 33 44

**Review by MuhanguziDavid — COMMENTED** [review]
Upon running the endpoint, a user is created. However, I do not get a token. I followed the steps in your PR description, could you have left something out?
Screenshot 2020-02-10 at 17 35 58

File: env.example [hunk 7]
```diff
@@ -3,4 +3,5 @@ DB_PASS='password of your postgress'
 DB_NAME='name of the database || Barefoot'
 DB_TEST_NAME='name of the database for test || Barefoot_test'
 NODE_ENV='development'
+secretkey='PrivateKey'
```
Comment (MuhanguziDavid) [inline env.example]: Make this upper case for consistency

File: src/routes/users.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+          {
+            "name": "gender",
+            "in": "formData",
+            "required": true,
+            "type": "string",
+            "description": ""
+          }
+        ],
+        "responses": {
+          "200": {
```
Comment (MuhanguziDavid) [inline src/routes/users.js]: the correct response should be 201

File: src/tests/index.test.js [hunk 9]
```diff
@@ -4,15 +4,16 @@ import chaiHttp from 'chai-http';
 import { describe, it } from 'mocha';
 import app from '../app';
 
+
 chai.use(chaiHttp);
 describe('SERVER CONFIG CHECK', () => {
-  it('Should check development error handler', (done) => {
+  it('it should return 404 when the provided rout does not exist', (done) => {
```
Comment (MuhanguziDavid) [inline src/tests/index.test.js]: please correct the spelling of `route`

**Review by MuhanguziDavid — APPROVED** [review]
LGTM

**Pull request merged**